add post about valgrind + gdb (#60)
reside-ic/reside-ic.github.io: content/blog/debugging-valgrind-gdb.md

---
title: "Debugging memory errors with valgrind and gdb"
author: Rich FitzJohn
date: 2021-06-10
tags:
 - dust
 - R
 - debugging
---

> We've already written about debugging R packages with valgrind and gdb together [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  This will likely only work on Linux (not windows or macOS).

Reviewer: ```suggestion
We've already written about debugging R packages with valgrind and gdb separately [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  NB: This will likely only work on Linux (not windows or macOS).
```

> We've already written about debugging R packages with valgrind and gdb together [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  This will likely only work on Linux (not windows or macOS).

Reviewer 2: ```suggestion
We've already written about debugging R packages with valgrind and gdb together [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  This will likely only work on Linux (not Windows or macOS).
```

> We've already written about debugging R packages with valgrind and gdb separately [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  NB: This will likely only work on Linux (not windows or macOS).

Reviewer 2: ```suggestion
We've already written about debugging R packages with valgrind and gdb separately [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  NB: This will likely only work on Linux (not Windows or macOS).
```

## The problem

> We noticed that as of 0.9.3 of `dust` we had a `std::bad_alloc` error when running models with the `float` type (rather than `double`). After a bit of narrowing it down we managed to reduce the problem to this bit of code:

Reviewer: Any easy way to explain how it was narrowed down? I think this is something that can be difficult in R

Author: Hmm, not much to add - I just took my crashing code and knocked out lines until I got it as simple as I wanted. The original code was from our benchmarking explorations. Or was there something else you meant specifically?

The seed thing is likely not surprising to people using R as it's easy enough to configure the seed there

Reviewer: Perhaps just that 'I just took my crashing code and knocked out lines until I got it as simple as I wanted'

> ```

Reviewer 2: ```suggestion
```r
```

sirs <- dust::dust_example("sirs")
pars <- list(beta = 0.2, gamma = 0.1, alpha = 0.1, freq = 4)
end <- 150 * 4
steps <- seq(0, end, by = 4)
ans <- sirs$new(pars, 0, 1, seed = 1L)$simulate(steps)
dat <- data.frame(step = steps[-1],
                  date = steps[-1] / 4,
                  incidence = ans[4, 1, -1])
dat_dust <- dust::dust_data(dat, "step")

path_sirs <- system.file("examples/sirs.cpp", package = "dust")
sirs_gpu <- dust::dust(path_sirs, gpu = FALSE, real_t = "float")

set.seed(1)

mod <- sirs_gpu$new(list(), 0, 100, seed = 42L)
mod$set_data(dat_dust)
for (i in 1:20) {
  message(i)
  mod$reset(list(), 0)
  mod$filter()
}
```

> This will crash deterministic ally on the 18th iteration through the loop.  Once we narrowed this down, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):

Reviewer: ```suggestion
This will crash deterministically on the 18th iteration through the loop.  Once we could reliable reproduce the error, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):
```

Reviewer: Don't know if it's worth saying something about the advantage of dust design, that the RNG seeding really helps avoid stochastic errors

> This will crash deterministically on the 18th iteration through the loop.  Once we could reliable reproduce the error, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):

Reviewer 2: ```suggestion
This will crash deterministically on the 18th iteration through the loop.  Once we could reliably reproduce the error, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):
```

==70753==    by 0x1437D28A: dust::Dust<sirs>::resample(std::vector<float, std::allocator<float> > const&, std::vector<unsigned long, std::allocator<unsigned long> >&) (dust.hpp:420)
==70753==    by 0x143802A3: std::vector<sirs::real_t, std::allocator<sirs::real_t> > dust::filter::filter<sirs>(dust::Dust<sirs>*, dust::filter::filter_state_host<sirs::real_t>&, bool, std::vector<unsigned long, std::allocator<unsigned long> >) (filter.hpp:59)
==70753==    by 0x14376629: cpp11::writable::r_vector<double> dust::r::run_filter<sirs, dust::filter::filter_state_host<float> >(dust::Dust<sirs>*, cpp11::sexp&, cpp11::sexp&, std::vector<unsigned long, std::allocator<unsigned long> >&, bool) (interface.hpp:442)
==70753==    by 0x1436C04C: cpp11::sexp dust::r::dust_filter<sirs, 0>(SEXPREC*, bool, cpp11::sexp, bool) (interface.hpp:475)
==70753==    by 0x14360D0B: dust_sirs_filter(SEXPREC*, bool, cpp11::sexp, bool) (dust.cpp:262)
==70753==    by 0x1435CAF6: _sirsbe1444bc_dust_sirs_filter (cpp11.cpp:117)
```

The top-left function there (`dust::filter::resample_weight`) looks like this:

> ```

Reviewer 2: ```suggestion
```cpp
```

template <typename real_t>
void resample_weight(typename std::vector<real_t>::const_iterator w,
                     size_t n, real_t u, size_t offset,
                     typename std::vector<size_t>::iterator idx) {
  const real_t tot = std::accumulate(w, w + n, static_cast<real_t>(0));
  real_t ww = 0.0, uu = tot * u / n, du = tot / n;

  size_t j = offset;
  for (size_t i = 0; i < n; ++i) {
    while (ww < uu) {
      ww += *w;
      ++w;
      ++j;
    }
    uu += du;
    *idx = j == 0 ? 0 : j - 1;
    ++idx;
  }
}
```

and line 20 is the line with `ww += *w;` (see [the source](https://github.com/mrc-ide/dust/blob/v0.9.3/inst/include/dust/filter_tools.hpp#L20]).

The problem was, it is not really obvious what is wrong at that point, or why it might be being triggered only when run with `float` version of that template.  We wanted to be able to inspect the values of variables at the point of the error.

> This is documented in [the valgrind docs](https://valgrind.org/docs/manual/manual-core-adv.html#manual-core-adv.gdbserver), and I'd done it a number of years ago.  First, start valgrind up with the argument `--vgdb-error=0` which starts some sort of server process that gdb can connect to, and sets it up to throw a signal on the first memory error

Reviewer 2: ```suggestion
This is documented in [the valgrind docs](https://valgrind.org/docs/manual/manual-core-adv.html#manual-core-adv.gdbserver), and I'd done it a number of years ago.  First, start valgrind up with the argument `--vgdb-error=0` which starts some sort of server process that gdb can connect to, and sets it up to throw a signal on the first memory error:
```

==72681==    by 0x143802A3: std::vector<sirs::real_t, std::allocator<sirs::real_t> > dust::filter::filter<sirs>(dust::Dust<sirs>*, dust::filter::filter_state_host<sirs::real_t>&, bool, std::vector<unsigned long, std::allocator<unsigned long> >) (filter.hpp:59)
==72681==    by 0x14376629: cpp11::writable::r_vector<double> dust::r::run_filter<sirs, dust::filter::filter_state_host<float> >(dust::Dust<sirs>*, cpp11::sexp&, cpp11::sexp&, std::vector<unsigned long, std::allocator<unsigned long> >&, bool) (interface.hpp:442)
[...snip...]
==72681==
==72681== (action on error) vgdb me ...
```

Control was then passed back to the second window running gdb:

```
Program received signal SIGTRAP, Trace/breakpoint trap.
dust::filter::resample_weight<float> (w=1.1479437e-41, n=100, u=0.999940574, offset=0, idx=99)
    at /home/rfitzjoh/lib/R/library/dust/include/dust/filter_tools.hpp:20
20	      ww += *w;
(gdb)
```

and all the usual gdb tricks work:

```
(gdb) print u
$1 = 0.999940574
(gdb) print du
$2 = 0.602285564
(gdb) print uu
$3 = 60.2285919
(gdb) print ww
$4 = 60.2285576
```

> Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `ww` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)

Reviewer 3: Do you mean this? Was confusing at first.

```suggestion
Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `w` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)
```

Reviewer 3: Oh, just looked at the PR. It was `j`

```suggestion
Eventually I worked out from this (and from pulling the values into a small standalone C++ program) that we were accumulating round-off error by incrementing `uu` by `du` each time, which could be avoided if we computed `uu` as `uu0 + i * du`.  I also added a fix to prevent us ever trying to move `j` past the end of the iterator. The fix is [here](https://github.com/mrc-ide/dust/pull/238)
```

Author: Ah, this is unclear. It's actually the `w` being iterated past its end which was raising the first error. Then `j` ends up with something that will break the index later and cascade through to cause a crash. I'll tidy this up tomorrow

> Usually with `valgrind` it's enough to see where the error occurs to find the memory issue. The approach here is useful when you need more information, and while awkward felt nice enough once I'd remembered how to do it. Hopefully next time I need to do this I'll remember this blog post exists.

Reviewer 2: For the benefit of someone who understands what valgrind is for but has never used it in anger (e.g. me!): what role did it actually play in this investigation? Presumably the original error occurs whether you're inside or outside valgrind... so I guess it's helping to locate the relevant code more precisely than using gdb alone?

Author: I've added a section on that now, good suggestion
